File: app/src/main/java/com/drdisagree/colorblendr/service/BroadcastListener.java
package com.drdisagree.colorblendr.service;

import static com.drdisagree.colorblendr.common.Const.FABRICATED_OVERLAY_NAME_APPS;
import static com.drdisagree.colorblendr.common.Const.MONET_LAST_UPDATED;
import static com.drdisagree.colorblendr.common.Const.MONET_SEED_COLOR;
import static com.drdisagree.colorblendr.common.Const.MONET_SEED_COLOR_ENABLED;
import static com.drdisagree.colorblendr.common.Const.SHIZUKU_THEMING_ENABLED;
import static com.drdisagree.colorblendr.common.Const.THEMING_ENABLED;
import static com.drdisagree.colorblendr.common.Const.WALLPAPER_COLOR_LIST;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.drdisagree.colorblendr.ColorBlendr;
import com.drdisagree.colorblendr.common.Const;
import com.drdisagree.colorblendr.config.RPrefs;
import com.drdisagree.colorblendr.extension.MethodInterface;
import com.drdisagree.colorblendr.provider.RootConnectionProvider;
import com.drdisagree.colorblendr.utils.AppUtil;
import com.drdisagree.colorblendr.utils.OverlayManager;
import com.drdisagree.colorblendr.utils.SystemUtil;
import com.drdisagree.colorblendr.utils.WallpaperUtil;

import java.util.ArrayList;
import java.util.HashMap;

public class BroadcastListener extends BroadcastReceiver {

    private static final String TAG = BroadcastListener.class.getSimpleName();
    public static int lastOrientation = -1;
    private static long cooldownTime = 5000;

    @SuppressWarnings("deprecation")
    @Override
    public void onReceive(Context context, Intent intent) {
        Log.d(TAG, "Received intent: " + intent.getAction());

        if (lastOrientation == -1) {
            lastOrientation = SystemUtil.getScreenRotation(context);
        }

        int currentOrientation = SystemUtil.getScreenRotation(context);

        if (Intent.ACTION_BOOT_COMPLETED.equals(intent.getAction()) ||
                Intent.ACTION_LOCKED_BOOT_COMPLETED.equals(intent.getAction())
        ) {
            // Start background service on boot
            if (AppUtil.permissionsGranted(context)) {
                if (BackgroundService.isServiceNotRunning()) {
                    context.startService(new Intent(ColorBlendr.getAppContext(), BackgroundService.class));
                }
            }

            validateRootAndUpdateColors(context, new MethodInterface() {
                @Override
                public void run() {
                    cooldownTime = 10000;
                    new Handler(Looper.getMainLooper()).postDelayed(() -> cooldownTime = 5000, 10000);
                    updateAllColors(context);
                }
            });
        }

        // Update wallpaper colors on wallpaper change
        if (Intent.ACTION_WALLPAPER_CHANGED.equals(intent.getAction()) &&
                AppUtil.permissionsGranted(context)
        ) {
            ArrayList<Integer> wallpaperColors = WallpaperUtil.getWallpaperColors(context);
            RPrefs.putString(WALLPAPER_COLOR_LIST, Const.GSON.toJson(wallpaperColors));

            if (!RPrefs.getBoolean(MONET_SEED_COLOR_ENABLED, false)) {
                RPrefs.putInt(MONET_SEED_COLOR, wallpaperColors.get(0));
            }
        }

        // Update fabricated colors on wallpaper change
        if (Intent.ACTION_WALLPAPER_CHANGED.equals(intent.getAction()) ||
                (Intent.ACTION_CONFIGURATION_CHANGED.equals(intent.getAction()) &&
                        lastOrientation == currentOrientation)
        ) {
            validateRootAndUpdateColors(context, new MethodInterface() {
                @Override
                public void run() {
                    updateAllColors(context);
                }
            });
        } else if (lastOrientation != currentOrientation) {
            lastOrientation = currentOrientation;
        }

        if (Intent.ACTION_PACKAGE_REMOVED.equals(intent.getAction())) {
            // Remove fabricated colors for uninstalled apps
            Uri data = intent.getData();

            if (data != null) {
                String packageName = data.getSchemeSpecificPart();
                HashMap<String, Boolean> selectedApps = Const.getSelectedFabricatedApps();

                if (selectedApps.containsKey(packageName) && Boolean.TRUE.equals(selectedApps.get(packageName))) {
                    selectedApps.remove(packageName);
                    Const.saveSelectedFabricatedApps(selectedApps);

                    validateRootAndUpdateColors(context, new MethodInterface() {
                        @Override
                        public void run() {
                            OverlayManager.unregisterFabricatedOverlay(String.format(FABRICATED_OVERLAY_NAME_APPS, packageName));
                        }
                    });
                }
            }
        } else if (Intent.ACTION_MY_PACKAGE_REPLACED.equals(intent.getAction())) {
            // Update fabricated colors for updated app
            validateRootAndUpdateColors(context, new MethodInterface() {
                @Override
                public void run() {
                    updateAllColors(context);
                }
            });
        } else if (Intent.ACTION_PACKAGE_REPLACED.equals(intent.getAction())) {
            // Update fabricated colors for updated app
            Uri data = intent.getData();

            if (data != null) {
                String packageName = data.getSchemeSpecificPart();
                HashMap<String, Boolean> selectedApps = Const.getSelectedFabricatedApps();

                if (selectedApps.containsKey(packageName) && Boolean.TRUE.equals(selectedApps.get(packageName))) {
                    validateRootAndUpdateColors(context, new MethodInterface() {
                        @Override
                        public void run() {
                            OverlayManager.applyFabricatedColorsPerApp(context, packageName, null);
                        }
                    });
                }
            }
        }

        if (Intent.ACTION_PACKAGE_ADDED.equals(intent.getAction()) ||
                Intent.ACTION_PACKAGE_REMOVED.equals(intent.getAction()) ||
                Intent.ACTION_WALLPAPER_CHANGED.equals(intent.getAction())) {
            LocalBroadcastManager.getInstance(context).sendBroadcast(intent);
        }
    }

    private static void validateRootAndUpdateColors(Context context, MethodInterface method) {
        if (Const.getWorkingMethod() == Const.WORK_METHOD.ROOT &&
                RootConnectionProvider.isNotConnected()
        ) {
            RootConnectionProvider.builder(context)
                    .runOnSuccess(method)
                    .run();
        } else {
            method.run();
        }
    }

    private static void updateAllColors(Context context) {
        if (!RPrefs.getBoolean(THEMING_ENABLED, true) && !RPrefs.getBoolean(SHIZUKU_THEMING_ENABLED, true)) {
            return;
        }

        if (Math.abs(RPrefs.getLong(MONET_LAST_UPDATED, 0) - System.currentTimeMillis()) >= cooldownTime) {
            RPrefs.putLong(MONET_LAST_UPDATED, System.currentTimeMillis());
            new Handler(Looper.getMainLooper()).postDelayed(() -> OverlayManager.applyFabricatedColors(context), 500);
        }
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/service/BroadcastListener.kt
package com.drdisagree.colorblendr.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.drdisagree.colorblendr.data.common.Constant
import com.drdisagree.colorblendr.data.common.Constant.FABRICATED_OVERLAY_NAME_APPS
import com.drdisagree.colorblendr.data.common.Utilities.customColorEnabled
import com.drdisagree.colorblendr.data.common.Utilities.getLastColorAppliedTimestamp
import com.drdisagree.colorblendr.data.common.Utilities.getSelectedFabricatedApps
import com.drdisagree.colorblendr.data.common.Utilities.getWallpaperColorJson
import com.drdisagree.colorblendr.data.common.Utilities.isRootMode
import com.drdisagree.colorblendr.data.common.Utilities.isRootOrShizukuUnknown
import com.drdisagree.colorblendr.data.common.Utilities.isShizukuThemingEnabled
import com.drdisagree.colorblendr.data.common.Utilities.isThemingEnabled
import com.drdisagree.colorblendr.data.common.Utilities.screenOffColorUpdateEnabled
import com.drdisagree.colorblendr.data.common.Utilities.setSeedColorValue
import com.drdisagree.colorblendr.data.common.Utilities.setSelectedFabricatedApps
import com.drdisagree.colorblendr.data.common.Utilities.setWallpaperColorJson
import com.drdisagree.colorblendr.data.common.Utilities.updateColorAppliedTimestamp
import com.drdisagree.colorblendr.provider.RootConnectionProvider
import com.drdisagree.colorblendr.utils.AppUtil.permissionsGranted
import com.drdisagree.colorblendr.utils.OverlayManager.applyFabricatedColors
import com.drdisagree.colorblendr.utils.OverlayManager.applyFabricatedColorsPerApp
import com.drdisagree.colorblendr.utils.OverlayManager.unregisterFabricatedOverlay
import com.drdisagree.colorblendr.utils.WallpaperColorUtil.getWallpaperColors
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.math.abs

class BroadcastListener : BroadcastReceiver() {

    private val handler = Handler(Looper.getMainLooper())
    private var sleepRunnable: Runnable? = null
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    @Suppress("deprecation")
    override fun onReceive(context: Context, intent: Intent) {
        Log.d(TAG, "Received intent: " + intent.action)

        if (isLastConfigInitialized.not()) {
            lastConfig = Configuration(context.resources.configuration)
        }

        CoroutineScope(Dispatchers.Main).launch {
            when (intent.action) {
                Intent.ACTION_BOOT_COMPLETED,
                Intent.ACTION_LOCKED_BOOT_COMPLETED -> {
                    handleBootCompleted(context)
                }

                Intent.ACTION_WALLPAPER_CHANGED -> {
                    handleWallpaperChanged(context, true)
                }

                Intent.ACTION_SCREEN_OFF -> {
                    if (screenOffColorUpdateEnabled()) {
                        sleepRunnable = Runnable {
                            coroutineScope.launch {
                                handleWallpaperChanged(context)
                            }
                        }
                        handler.postDelayed(sleepRunnable!!, 15000) // 15 seconds
                    }
                }

                Intent.ACTION_SCREEN_ON -> {
                    sleepRunnable?.let { runnable ->
                        handler.removeCallbacks(runnable)
                        sleepRunnable = null
                    }
                }

                Intent.ACTION_CONFIGURATION_CHANGED -> {
                    val newConfig = Configuration(context.resources.configuration)
                    val lastUiMode = lastConfig.uiMode and Configuration.UI_MODE_NIGHT_MASK
                    val newUiMode = newConfig.uiMode and Configuration.UI_MODE_NIGHT_MASK

                    if (lastUiMode != newUiMode) {
                        delay(1000)
                        validateRootAndUpdateColors(context) {
                            updateAllColors()
                        }
                    }

                    lastConfig = newConfig
                }

                Intent.ACTION_PACKAGE_REMOVED -> {
                    handlePackageRemoved(context, intent)
                }

                Intent.ACTION_MY_PACKAGE_REPLACED,
                Intent.ACTION_PACKAGE_REPLACED -> {
                    handlePackageReplaced(context, intent)
                }
            }

            if (intent.action in listOf(
                    Intent.ACTION_PACKAGE_ADDED,
                    Intent.ACTION_PACKAGE_REMOVED,
                    Intent.ACTION_WALLPAPER_CHANGED
                )
            ) {
                LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
            }
        }
    }

    private suspend fun handleBootCompleted(context: Context) {
        if (permissionsGranted(context) && AutoStartService.isServiceNotRunning) {
            context.startForegroundService(Intent(context, AutoStartService::class.java))
        }

        validateRootAndUpdateColors(context) {
            cooldownTime = 10000
            CoroutineScope(Dispatchers.Main).launch {
                delay(10000)
                cooldownTime = 5000
            }
            updateAllColors()
        }
    }

    private suspend fun handleWallpaperChanged(context: Context, force: Boolean = false) {
        if (permissionsGranted(context)) {
            val wallpaperColors = withContext(Dispatchers.IO) {
                getWallpaperColors(context)
            }

            val previousWallpaperColors = getWallpaperColorJson()
            val currentWallpaperColors = Constant.GSON.toJson(wallpaperColors)

            if (!requiresUpdate) {
                requiresUpdate = previousWallpaperColors != currentWallpaperColors
            }

            setWallpaperColorJson(currentWallpaperColors)

            if (!customColorEnabled()) {
                setSeedColorValue(wallpaperColors[0])
            }
        }

        if (requiresUpdate || force) {
            requiresUpdate = false
            validateRootAndUpdateColors(context) {
                updateAllColors()
            }
        }
    }

    private suspend fun handlePackageRemoved(context: Context, intent: Intent) {
        intent.data?.schemeSpecificPart?.let { packageName ->
            val selectedApps: HashMap<String, Boolean> = getSelectedFabricatedApps()

            if (selectedApps.containsKey(packageName) && selectedApps[packageName] == true) {
                selectedApps.remove(packageName)
                setSelectedFabricatedApps(selectedApps)

                validateRootAndUpdateColors(context) {
                    unregisterFabricatedOverlay(
                        String.format(
                            FABRICATED_OVERLAY_NAME_APPS,
                            packageName
                        )
                    )
                }
            }
        }
    }

    private suspend fun handlePackageReplaced(context: Context, intent: Intent) {
        intent.data?.schemeSpecificPart?.let { packageName ->
            val selectedApps: HashMap<String, Boolean> = getSelectedFabricatedApps()

            if (selectedApps.containsKey(packageName) && selectedApps[packageName] == true) {
                validateRootAndUpdateColors(context) {
                    applyFabricatedColorsPerApp(packageName, null)
                }
            }
        }
    }

    private suspend fun validateRootAndUpdateColors(context: Context, method: suspend () -> Unit) {
        if (isRootMode() && RootConnectionProvider.isNotConnected) {
            RootConnectionProvider
                .builder(context)
                .onSuccess {
                    CoroutineScope(Dispatchers.Main).launch {
                        method()
                    }
                }
                .run()
        } else {
            method()
        }
    }

    private fun updateAllColors() {
        if ((!isThemingEnabled() && !isShizukuThemingEnabled()) || isRootOrShizukuUnknown()) return

        if (abs(System.currentTimeMillis() - getLastColorAppliedTimestamp()) >= cooldownTime) {
            updateColorAppliedTimestamp()

            CoroutineScope(Dispatchers.Main).launch {
                delay(500)
                applyFabricatedColors()
            }
        }
    }

    companion object {
        private val TAG: String = BroadcastListener::class.java.simpleName
        var requiresUpdate = false
        private var cooldownTime: Long = 5000
        lateinit var lastConfig: Configuration
        val isLastConfigInitialized: Boolean
            get() = ::lastConfig.isInitialized
    }
}
